pkg/api/devops/v1alpha4/pipelinerun_types.go

```diff
@@ -0,0 +1,199 @@
+
+	// Completion timestamp of the pipeline run.
+	// +optional
+	CompletionTime *metav1.Time `json:"completionTime,omitempty"`
+
+	// Update timestamp of the pipeline run.
+	// +optional
+	UpdateTime *metav1.Time `json:"updateTime,omitempty"`
+
+	// Current state of pipeline run.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions []Condition `json:"conditions,omitempty"`
+
+	// Current phase of pipeline run.
+	// +optional
+	Phase RunPhase `json:"phase,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// PipelineRun is the Schema for the pipelineruns API
+type PipelineRun struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   PipelineRunSpec   `json:"spec,omitempty"`
+	Status PipelineRunStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// PipelineRunList contains a list of PipelineRun
+type PipelineRunList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []PipelineRun `json:"items"`
+}
+
+// Parameter is an option that can be passed with the endpoint to influence the Pipeline Run
+type Parameter struct {
+
+	// Name indicates that name of the parameter.
+	Name string `json:"name" description:"parameter name"`
+
+	// Value indicates that value of the parameter.
+	Value string `json:"value" description:"parameter value"`
+}
+
+// RefType indicates that scm reference type, such as branch, tag, pr, mr.
+type RefType string
+
+const (
+	// Branch represents an independent line of development.
+	Branch RefType = "branch"
+
+	// Tag points to specific points in SCM history.
+	Tag RefType = "tag"
+
+	// PullRequest is a feature that makes it easier for developers to collaborate using GitHub, BitBucket or others.
+	PullRequest RefType = "pr"
+
+	// MergeRequest is a request from someone to merge in code from one branch to another
+	MergeRequest RefType = "mr"
+)
+
+// Scm is a SCM configuration that target pipeline run requires.
+type Scm struct {
+	// RefType indicates that SCM reference type, such as branch, tag, pr, mr.
+	RefType RefType `json:"refType" description:"scm reference type"`
+
+	// RefName indicates that SCM reference name, such as master, dev, release-v1.
+	RefName string `json:"refName" description:"scm reference name"`
+}
+
+// RunPhase is a label for the condition of a pipeline run at the current time.
+type RunPhase string
+
+const (
+	// Pending indicates that the pipeline run is pending.
+	Pending RunPhase = "Pending"
+
+	// Running indicates that the pipeline run is running.
+	Running RunPhase = "Running"
+
+	// Succeeded indicates that the pipeline run has succeeded.
+	Succeeded RunPhase = "Succeeded"
+
+	// Failed indicates that the pipeline run has failed.
+	Failed RunPhase = "Failed"
+
+	// Unknown indicates that the pipeline run has an unknown status.
+	Unknown RunPhase = "Unknown"
+)
```

Comment:
These constants do not necessary in a versioned package. For example, we can put them in `pkg/api/utils.go`

Reply:
According to:
- https://github.com/kubernetes/api/blob/master/apps/v1/types.go
- https://github.com/kubernetes/api/blob/master/apps/v1beta1/types.go

I think those constants is reasonable in a versioned package.

Reply:
Fair enough.